Move Opaque Origin Identifier from SecurityOrigin to SecurityOriginData

sysrqb opened the pull request:
#### 7fde32c36b521c0b98efa92eb57a635018f8355f
<pre>
Move Opaque Origin Identifier from SecurityOrigin to SecurityOriginData
<a href="https://bugs.webkit.org/show_bug.cgi?id=251048">https://bugs.webkit.org/show_bug.cgi?id=251048</a>
rdar://104578586

Reviewed by Youenn Fablet, Sihui Liu and Chris Dumez.

This change is a pre-patch for later work. The motivation for this change is
that SecurityOriginData will be used in a HashMap and the key in that mapping
must correctly take the opaqueOriginIdentifier into account.

There are some unintended side-effects from this change, so we make some
relevant changes, as well. We modify some permissions checks in WebKit&apos;s
UIProcess and NetworkProcess because the expectations of the existing logic
fall into two categories:
  - converting a SecurityOriginData into a SecurityOrigin should consider
    creating an opaque origin
  - converting a URL into a SecurityOriginData should not consider creating an
    opaque origin.

Now with this patch neither of these actions are true. In the first case, now
we instantiate a SecurityOrigin directly with a SecurityOriginData, and in that
constructor we don&apos;t re-evaluate if an opaque origin should be created.
Therefore, that check must happen when creating the SecurityOriginData. In the
other case, we consider creating an opaque SecurityOriginData from a URL
instead of directly mapping the URL&apos;s scheme+host+port to the same fields in
the SecurityOriginData.

As a result of this, we now directly compare the scheme, host, and port of URLs
in some places instead of relying on SecurityOrigin::isSameSchemeHostPort. For
other cases, we introduce a new function
SecurityOriginData::fromURLWithoutStrictOpaqueness where we skip most Opaque
Origin checks.

We also explicitly set the URL in WebKitGLib&apos;s MediaStream tests because they
use about:blank as the default and that creates a discrepancy between the
Page&apos;s active URL and the top document&apos;s SecurityOrigin.

* Source/WTF/wtf/Markable.h:
(WTF::add):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::initializeShared):
(WebCore::SecurityOrigin::SecurityOrigin):
(WebCore::SecurityOrigin::create):
(WebCore::SecurityOrigin::isSameOriginDomain const):
(WebCore::SecurityOrigin::isSameOriginAs const):
(WebCore::SecurityOrigin::equal const):
(WebCore::schemeRequiresHost): Deleted.
(WebCore::shouldTreatAsOpaqueOrigin): Deleted.
* Source/WebCore/page/SecurityOrigin.h:
(WebCore::SecurityOrigin::isOpaque const):
(WebCore::add):
* Source/WebCore/page/SecurityOriginData.cpp:
(WebCore::SecurityOriginData::fromURL):
(WebCore::SecurityOriginData::fromURLWithoutStrictOpaqueness):
(WebCore::SecurityOriginData::securityOrigin const):
(WebCore::SecurityOriginData::isolatedCopy const):
(WebCore::SecurityOriginData::isolatedCopy):
(WebCore::operator==):
(WebCore::schemeRequiresHost):
(WebCore::SecurityOriginData::shouldTreatAsOpaqueOrigin):
* Source/WebCore/page/SecurityOriginData.h:
(WebCore::SecurityOriginData::SecurityOriginData):
(WebCore::SecurityOriginData::createOpaque):
(WebCore::SecurityOriginData::isOpaque const):
(WebCore::SecurityOriginData::encode const):
(WebCore::SecurityOriginData::decode):
(WebCore::add):
(WebCore::SecurityOriginData::fromURL): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataForOrigin):
(WebKit::NetworkProcess::renameOriginInWebsiteData):
(WebKit::NetworkProcess::websiteDataOriginDirectoryForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::createFetchTask):
(WebKit::WebSWServerConnection::registerServiceWorkerClient):
* Source/WebKit/Shared/API/APISecurityOrigin.h:
(API::SecurityOrigin::createFromString):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _renameOrigin:to:forDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _originDirectoryForTesting:topOrigin:type:completionHandler:]):
* Source/WebKit/UIProcess/API/glib/WebKitSecurityOrigin.cpp:
(webkit_security_origin_new_for_uri):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::renameOriginInWebsiteData):
(WebKit::NetworkProcessProxy::websiteDataOriginDirectoryForTesting):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::cancel):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::decidePolicyForUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::checkUserMediaPermissionForSpeechRecognition):
(WebKit::UserMediaPermissionRequestManagerProxy::shouldChangeDeniedToPromptForCamera const):
(WebKit::UserMediaPermissionRequestManagerProxy::shouldChangeDeniedToPromptForMicrophone const):
(WebKit::UserMediaPermissionRequestManagerProxy::getUserMediaPermissionInfo):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::didChangeMainDocument):
(WebKit::WebPageProxy::makeStorageSpaceRequest):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
(WebKit::WebPermissionControllerProxy::mostReasonableWebPageProxy const):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::forWebPagesWithOrigin):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::renameOriginInWebsiteData):
(WebKit::WebsiteDataStore::originDirectoryForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/ios/WKGeolocationProviderIOS.mm:
(-[WKGeolocationProviderIOS geolocationAuthorizationGranted]):
* Tools/TestWebKitAPI/Tests/WebCore/SecurityOrigin.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
(testWebViewUserMediaEnumerateDevicesPermissionCheck):
(testWebViewUserMediaPermissionRequests):
(testWebViewAudioOnlyUserMediaPermissionRequests):
(testWebViewDisplayUserMediaPermissionRequests):

Canonical link: <a href="https://commits.webkit.org/260214@main">https://commits.webkit.org/260214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbcb0ad99921a7d658823c93a588966428d69224

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16204 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116331 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115809 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17664 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7404 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99337 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112921 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13346 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96403 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40955 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95246 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28039 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82721 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/cut-copy-paste/editable (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96584 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9289 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29380 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96020 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7261 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9900 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6375 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30755 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15447 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48959 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104835 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7060 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11442 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25979 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -90,19 +103,58 @@ struct SecurityOriginData {
+    return SecurityOriginData(WTFMove(*protocol), WTFMove(*host), *port);
+}
+
+
```
line unneeded

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -42,19 +49,24 @@ struct SecurityOriginData {
-            url.host().isNull() ? emptyString() : url.host().convertToASCIILowercase(),
-            url.port()
-        };
+        return SecurityOriginData { SecurityOriginData::OpaqueOriginIdentifier::generateThreadSafe() };
```
nit: Do we really need the `SecurityOriginData::` here?

cdumez reviewed (approved):
LGTM if the bots are happy.

szewai commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -2420,7 +2420,15 @@ void NetworkProcess::websiteDataOriginDirectoryForTesting(PAL::SessionID session
+        };
+    };
+
+    auto clientOrigin = WebCore::ClientOrigin { fromURL(topOrigin), fromURL(origin) };
```
I think we should change the SPI `_originDirectoryForTesting` to take `WKSecurityOrigin` instead of `NSURL`, and use SecurityOriginData in related functions (e.g. `NetworkProcessProxy::websiteDataOriginDirectoryForTesting()`), instead of adding special check here.

To reduce the risk of breakage, I think we might add another `SecurityOriginData::fromURL(ShouldSkipOpaqueOriginCheck)` for places that currently use `SecurityOriginData::fromURL` for converting `URL` to `SecurityOriginData`.  For example the SPI `_renameOrigin` has similar pattern as `_originDirectoryForTesting`, but we cannot just change the interface to use `WKSecurityOrigin` without notifying SPI client.

sysrqb replied:
I agree that we should use `WKSecurityOrigin`, but it is difficult because all of the methods available for constructing a WKSecurityOrigin (`WKSecurityOriginCreate*`) wrap around `WebCore::SecurityOrigin::create*` and those `SecurityOrigin::create` functions includes opaque origin checks (which we want to avoid).

I added a new converting constructor for `SecurityOriginData`, where that constructor directly converts `URL->SecurityOriginData`. `SecurityOriginData::fromURL` checks whether the URL should create an opaque SecurityOriginData. I believe I audited all of the `SecurityOriginData::fromURL` call sites and made appropriate changes to either directly compare URLs with `protocolHostAndPortAreEqual` or construct `SecurityOriginData(URL)`.

szewai commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -90,19 +112,57 @@ struct SecurityOriginData {
 inline void add(Hasher& hasher, const SecurityOriginData& data)
 {
-    add(hasher, data.protocol, data.host, data.port);
+    add(hasher, data.protocol, data.host, data.port, data.opaqueOriginIdentifier);
```
This seems a behavior change and requires a test (as in Tools/TestWebKitAPI/Tests/WebCore/SecurityOrigin.cpp)
Two SecurityOrigins created with default constructor will have different hash now.

cdumez replied:
I agree this is a behavior change and it'd be good to add a test. I'd add that it seems like a progression to me. Distinct opaque origins *should* have different hashes.

szewai commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -42,19 +49,33 @@ struct SecurityOriginData {
-    static SecurityOriginData fromURL(const URL& url)
+    WEBCORE_EXPORT static SecurityOriginData fromURL(const URL&);
+
+    static SecurityOriginData createOpaque()
```
Why not just add a default constructor and make it return an object with opaqueOriginIdentifier, as `SecurityOrigin()`?
And it seems wrong if we don't make `SecurityOriginData()` return object with an opaque identifier, as the object will have empty url fields while `isOpaque()` return false.

cdumez replied:
SecurityOriginData distinguishes the null state from the opaque state. The default constructor sets the strings to null, not the empty string. See this existing code:

```
    bool isNull() const
    {
        return protocol.isNull() && host.isNull() && port == std::nullopt;
    }
    bool isOpaque() const
    {
        return protocol == emptyString() && host == emptyString() && !port;
    }
```

As a result, I do not agree with your comment.

szewai replied:
For a SecurityOriginData created from default constructor, isOpaque() returns true before this patch, and returns false after this patch (which also modifies `isOpaque()`). Is this expected?

cdumez replied:
I don't think isOpaque() returned true before this patch for default-constructed SecurityOriginData?

The protocol & host would be null, not the empty string and thus would fail the isOpaque() check, no?

szewai replied:
oh right, I misread this. It's actually three states:
1. url fields null, opaqueID invalid
2. url fields empty, opaqueID valid
3. url fields non-empty, opaqueID invalid

szewai commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -91,7 +91,7 @@ static RefPtr<SecurityOrigin> getCachedOrigin(const URL& url)
 }
 
-static bool shouldTreatAsOpaqueOrigin(const URL& url)
+bool SecurityOrigin::shouldTreatAsOpaqueOrigin(const URL& url)
```
This should probably be moved to `SecurityOriginData`, instead of making it member of `SecurityOrigin`.

sysrqb replied:
Okay, I'll move. I moved it previously, but then moved it back and left it in `SecurityOrigin` because:
 - it relies on the static function `schemeRequiresHost()`, so that will move, too
 - it's used in `SecurityOrigin::SecurityOrigin(const URL&)`, too
 - I was avoiding adding a dependency on `LegacySchemeRegistry` to `SecurityOriginData`
 
 but none of these prevent moving it to `SecurityOriginData`.

szewai commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -42,19 +49,33 @@ struct SecurityOriginData {
+            , url.port()
+        }
+    {
+    }
```
My understanding is "empty url fields" and "empty opaqueOriginIdentifier" should be mutually exclusive, so `SecurityOriginData(const String& protocol, const String& host, std::optional<uint16_t> port)` should set `opaqueOriginIdentifier` if these fields are all empty.
The exclusion should be checked in other constructors.

Also, I think it might be better to add a flag to `fromURL()` to indicate whether the url should be checked for opaque origin (and assert url fields non null if no check), instead of adding a constructor. Because it's hard to tell the difference by looking at the header file

sysrqb replied:
Good point, this constructor should create an opaque origin if the URL is null. I'll fix that. And regarding constructor verses `fromURL()`, I initially added the flag to `fromURL()` but I realized creating a constructor seemed more natural in this case, so I change it. But, choosing the correct method relies on some subtlety, so I'll go back to using the flag since that seems better and more explicit.

szewai commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -1621,12 +1621,6 @@ struct WebCore::DataListSuggestionInformation {
 };
 #endif
 
-struct WebCore::SecurityOriginData {
-    [Validator='!protocol->isHashTableDeletedValue()'] String protocol;
-    String host;
-    std::optional<uint16_t> port;
-};
-
```
Why is this moved?

cdumez replied:
See comments above.

cdumez replied:
I am having trouble findings earlier review feedback in the Github UI. If I remember correctly, the issue is that this required a SecurityOriginData constructor which took [protocol, host, port, opaqueIdentifier] and such constructor didn't make much sense since you should never be passing a [protocol, host, port] when you have an opaqueIdentifier.

I wanted 2 constructors, one that takes a [protocol, host, port] and another that takes an opaqueIdentifier.

sysrqb replied:
That was in the previous PR: https://github.com/WebKit/WebKit/pull/9000#discussion_r1084786582 and https://github.com/WebKit/WebKit/pull/9000#discussion_r1084710188

cdumez commented on Source/WebCore/page/SecurityOriginData.cpp:
```diff
@@ -66,9 +67,23 @@ SecurityOriginData SecurityOriginData::fromFrame(Frame* frame)
 
+SecurityOriginData SecurityOriginData::fromURL(const URL& url, ShouldSkipOpaqueSecurityOriginCheck shouldSkipOpaqueSecurityOriginCheck)
+{
+    bool shouldCreateOpaqueOrigin = ShouldSkipOpaqueSecurityOriginCheck::No == shouldSkipOpaqueSecurityOriginCheck && shouldTreatAsOpaqueOrigin(url);
```
I personally think that `shouldSkipOpaqueSecurityOriginCheck  == ShouldSkipOpaqueSecurityOriginCheck::No` looks better than `ShouldSkipOpaqueSecurityOriginCheck::No == shouldSkipOpaqueSecurityOriginCheck`.

szewai commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -176,7 +128,14 @@ bool shouldTreatAsPotentiallyTrustworthy(const URL& url)
+        m_filePath = url.fileSystemPath(); // In case enforceFilePathSeparation() is called.
+}
+
+SecurityOrigin::SecurityOrigin(SecurityOriginData&& data)
```
It seems `m_canLoadLocalResources` might not be set for a SecurityOrigin object created by `SecurityOrigin::create(SecurityOriginData&& data)` because `SecurityOrigin::SecurityOrigin(SecurityOriginData&& data)` does not set it any more.

sysrqb replied:
I'm assuming you meant `m_filePath`, instead of `m_canLoadLocalResources` (`m_canLoadLocalResources` is set by `m_canLoadLocalResources = isLocal();`).

I believe you're correct that we should set `m_filePath` in `SecurityOrigin::SecurityOrigin(SecurityOriginData&&)`, too. I'll fix that.

szewai reviewed (commented):
r=me if the comment is resolved

cdumez commented on Source/WebCore/page/SecurityOriginData.cpp:
```diff
@@ -66,9 +67,23 @@ SecurityOriginData SecurityOriginData::fromFrame(Frame* frame)
+
+    if (shouldCreateOpaqueOrigin || (url.protocol().isNull() && url.host().isNull()))
+        return createOpaque();
+    return SecurityOriginData {
```
Shouldn't we create an opaque URL if url.protocol() is null/empty and url.host() is null/empty?
Right now, it seems we'll construct a SecurityOriginData with protocol&host being the emptyString, which used to be recognized as opaque, but not anymore after your change?

sysrqb replied:
Yes, good catch. I moved the empty/null checks into the constructor. It's not very pretty, so happy to hear suggestions.

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -42,19 +51,24 @@ struct SecurityOriginData {
+    explicit SecurityOriginData(OpaqueOriginIdentifier opaqueOriginIdentifier)
+        : protocol(emptyString())
+        , host(emptyString())
+        , port(std::nullopt)
```
Not needed.

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -25,15 +25,24 @@
 class Frame;
 class SecurityOrigin;
 
+enum class ShouldSkipOpaqueSecurityOriginCheck : bool { No, Yes };
```
I don't have a better proposal right now but this is ugly and so many call sites have to pass this parameter now..
It makes the new code look strictly worse (more verbose), which is unfortunate.

cdumez replied:
Maybe instead of an enum parameter, we could use an overload? I feel that we could fine a name short enough that it would result in more concise code. E.g. `fromURL()` and `fromURLIgnoringOpaqueness()`? or `fromNonOpaqueURL()` ?

sysrqb replied:
I used `fromURLIgnoringOpaqueness`. I also considered:
 - `fromURLToTupleOrigin`
 - `fromURLWithOpaquenessUnchecked`
 - `fromURLWithoutOpaqueness`
 - `directlyFromURL`
 - `fromURLToEquivalent`
 
 Plus some other similar names.

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -25,36 +25,52 @@
         , port(port)
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
+        if ((protocol.isEmpty() || protocol.isNull()) && (host.isEmpty() || host.isNull()))
```
a null SecurityOriginData is different from an opaque SecurityOriginData. Therefore, setting *this to an opaque origin  when protocol && host are null seems wrong to me.

cdumez replied:
Also, seems we should check that `!port` to maintain pre-existing behavior.

sysrqb replied:
The `null` protocol/host behavior was based on your comment in https://github.com/WebKit/WebKit/pull/9253#discussion_r1099321242 and also https://github.com/WebKit/WebKit/pull/9253#discussion_r1096352576. But, I suppose we should allow constructing a null SecurityOriginData through this constructor. I was avoiding duplicating opaqueness checks in this constructor and `fromURL`, but I think we do need them in both places.

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -25,36 +25,52 @@
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
+        if ((protocol.isEmpty() || protocol.isNull()) && (host.isEmpty() || host.isNull()))
+            *this = createOpaque();
```
Why can't we just initialize `opaqueOriginIdentifier`?

sysrqb replied:
If protocol/host are emptyString(), then we should re-set them null, and calling `createOpaque` seemed cleaner. But, given my other response, I'll be changing this.

cdumez commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -153,7 +153,7 @@ void ProvisionalPageProxy::cancel()
         true, // isMainFrame
         m_request,
-        SecurityOriginData::fromURL(m_request.url()),
+        page().urlSchemeHandlerForScheme(m_request.url().protocol().toStringWithoutCopying()) ? SecurityOriginData::fromURLIgnoringOpaqueness(m_request.url()) : SecurityOriginData::fromURL(m_request.url()),
```
Why can't we use `SecurityOriginData::fromURLIgnoringOpaqueness()` unconditionally here?

sysrqb replied:
I discussed this with Alex, and he made the argument that looking at the current registered custom schemes (when it's available outside of the web process) was safer than unconditionally ignoring the opaqueness checks. This is based on the observation that the primary reason why calling `fromURL` in the UIProcess results in an opaque origin is because of the difference in registered scheme handlers within the LegacySchemeRegistry.

cdumez replied:
But we were never doing an opaqueness check here so you're changing behavior. The new code is more complicated and the change is risky.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3583,7 +3583,9 @@ void WebPageProxy::receivedNavigationPolicyDecision(PolicyAction policyAction, A
     if (navigation && (!navigation->websitePolicies() || navigation->websitePolicies()->deviceOrientationAndMotionAccessState() == WebCore::DeviceOrientationOrMotionPermissionState::Prompt)) {
-        auto deviceOrientationPermission = websiteDataStore->deviceOrientationAndMotionAccessController().cachedDeviceOrientationPermission(SecurityOriginData::fromURL(navigation->currentRequest().url()));
+        auto& url = navigation->currentRequest().url();
+        auto origin = urlSchemeHandlerForScheme(url.protocol().toStringWithoutCopying()) ? SecurityOriginData::fromURLIgnoringOpaqueness(url) : SecurityOriginData::fromURL(url);
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5482,7 +5484,7 @@ void WebPageProxy::didChangeMainDocument(FrameIdentifier frameID)
         if (auto* gpuProcess = m_process->processPool().gpuProcess()) {
             if (auto* frame = WebFrameProxy::webFrame(frameID))
-                gpuProcess->updateCaptureOrigin(SecurityOriginData::fromURL(frame->url()), m_process->coreProcessIdentifier());
+                gpuProcess->updateCaptureOrigin(urlSchemeHandlerForScheme(frame->url().protocol().toStringWithoutCopying()) ? SecurityOriginData::fromURLIgnoringOpaqueness(frame->url()) : SecurityOriginData::fromURL(frame->url()), m_process->coreProcessIdentifier());
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8972,7 +8974,7 @@ void WebPageProxy::makeStorageSpaceRequest(FrameIdentifier frameID, const String
 
     auto originData = SecurityOriginData::fromDatabaseIdentifier(originIdentifier);
-    if (originData != SecurityOriginData::fromURL(URL { currentURL() })) {
+    if (!originData || !protocolHostAndPortAreEqual(originData->toURL(), URL { currentURL() })) {
```
Why is this better than relying on `SecurityOriginData::fromURLIgnoringOpaqueness()`?

sysrqb replied:
This came from a previous version when we believe we needed `fromURLIgnoringOpaqueness`. We could use `fromURLIgnoringOpaqueness` in these places, but we also avoid any issues by comparing URLs instead of `SecurityOriginData`s. If constructing and comparing `SecurityOriginData` in these checks is better, then I'll make that change.

cdumez replied:
It would seem like a more minimal change

cdumez commented on Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
```diff
@@ -77,7 +77,7 @@ RefPtr<WebPageProxy> WebPermissionControllerProxy::mostReasonableWebPageProxy(co
                 if (!potentialWebPageProxy)
                     continue;
-                if (WebCore::SecurityOriginData::fromURL(URL { potentialWebPageProxy->currentURL() }) != topOrigin)
+                if (!protocolHostAndPortAreEqual(URL { potentialWebPageProxy->currentURL() }, topOrigin.toURL()))
```
There are quite a few places like this why I don't understand why we didn't simply use `SecurityOriginData::fromURLIgnoringOpaqueness()` instead of `protocolHostAndPortAreEqual()`.

cdumez reviewed (approved):
LGTM

achristensen07 commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -90,19 +103,59 @@ struct SecurityOriginData {
 };
 
+template<class Encoder>
+void SecurityOriginData::encode(Encoder& encoder) const
```
Moving from WebCoreArgumentCoders.serialization.in to here is a step in the wrong direction.  This is what the ReturnEarlyIfTrue keyword is for.

sysrqb replied:
Fun, okay. I'll fix that in a follow-up, unless you want to block merging on this.

sysrqb committed:
Move Opaque Origin Identifier from SecurityOrigin to SecurityOriginData
https://bugs.webkit.org/show_bug.cgi?id=251048
rdar://104578586

Reviewed by Youenn Fablet, Sihui Liu and Chris Dumez.

This change is a pre-patch for later work. The motivation for this change is
that SecurityOriginData will be used in a HashMap and the key in that mapping
must correctly take the opaqueOriginIdentifier into account.

There are some unintended side-effects from this change, so we make some
relevant changes, as well. We modify some permissions checks in WebKit's
UIProcess and NetworkProcess because the expectations of the existing logic
fall into two categories:
  - converting a SecurityOriginData into a SecurityOrigin should consider
    creating an opaque origin
  - converting a URL into a SecurityOriginData should not consider creating an
    opaque origin.

Now with this patch neither of these actions are true. In the first case, now
we instantiate a SecurityOrigin directly with a SecurityOriginData, and in that
constructor we don't re-evaluate if an opaque origin should be created.
Therefore, that check must happen when creating the SecurityOriginData. In the
other case, we consider creating an opaque SecurityOriginData from a URL
instead of directly mapping the URL's scheme+host+port to the same fields in
the SecurityOriginData.

As a result of this, we now directly compare the scheme, host, and port of URLs
in some places instead of relying on SecurityOrigin::isSameSchemeHostPort. For
other cases, we introduce a new function
SecurityOriginData::fromURLWithoutStrictOpaqueness where we skip most Opaque
Origin checks.

We also explicitly set the URL in WebKitGLib's MediaStream tests because they
use about:blank as the default and that creates a discrepancy between the
Page's active URL and the top document's SecurityOrigin.

* Source/WTF/wtf/Markable.h:
(WTF::add):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::initializeShared):
(WebCore::SecurityOrigin::SecurityOrigin):
(WebCore::SecurityOrigin::create):
(WebCore::SecurityOrigin::isSameOriginDomain const):
(WebCore::SecurityOrigin::isSameOriginAs const):
(WebCore::SecurityOrigin::equal const):
(WebCore::schemeRequiresHost): Deleted.
(WebCore::shouldTreatAsOpaqueOrigin): Deleted.
* Source/WebCore/page/SecurityOrigin.h:
(WebCore::SecurityOrigin::isOpaque const):
(WebCore::add):
* Source/WebCore/page/SecurityOriginData.cpp:
(WebCore::SecurityOriginData::fromURL):
(WebCore::SecurityOriginData::fromURLWithoutStrictOpaqueness):
(WebCore::SecurityOriginData::securityOrigin const):
(WebCore::SecurityOriginData::isolatedCopy const):
(WebCore::SecurityOriginData::isolatedCopy):
(WebCore::operator==):
(WebCore::schemeRequiresHost):
(WebCore::SecurityOriginData::shouldTreatAsOpaqueOrigin):
* Source/WebCore/page/SecurityOriginData.h:
(WebCore::SecurityOriginData::SecurityOriginData):
(WebCore::SecurityOriginData::createOpaque):
(WebCore::SecurityOriginData::isOpaque const):
(WebCore::SecurityOriginData::encode const):
(WebCore::SecurityOriginData::decode):
(WebCore::add):
(WebCore::SecurityOriginData::fromURL): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataForOrigin):
(WebKit::NetworkProcess::renameOriginInWebsiteData):
(WebKit::NetworkProcess::websiteDataOriginDirectoryForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::createFetchTask):
(WebKit::WebSWServerConnection::registerServiceWorkerClient):
* Source/WebKit/Shared/API/APISecurityOrigin.h:
(API::SecurityOrigin::createFromString):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _renameOrigin:to:forDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _originDirectoryForTesting:topOrigin:type:completionHandler:]):
* Source/WebKit/UIProcess/API/glib/WebKitSecurityOrigin.cpp:
(webkit_security_origin_new_for_uri):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::renameOriginInWebsiteData):
(WebKit::NetworkProcessProxy::websiteDataOriginDirectoryForTesting):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::cancel):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::decidePolicyForUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::checkUserMediaPermissionForSpeechRecognition):
(WebKit::UserMediaPermissionRequestManagerProxy::shouldChangeDeniedToPromptForCamera const):
(WebKit::UserMediaPermissionRequestManagerProxy::shouldChangeDeniedToPromptForMicrophone const):
(WebKit::UserMediaPermissionRequestManagerProxy::getUserMediaPermissionInfo):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::didChangeMainDocument):
(WebKit::WebPageProxy::makeStorageSpaceRequest):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
(WebKit::WebPermissionControllerProxy::mostReasonableWebPageProxy const):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::forWebPagesWithOrigin):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::renameOriginInWebsiteData):
(WebKit::WebsiteDataStore::originDirectoryForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/ios/WKGeolocationProviderIOS.mm:
(-[WKGeolocationProviderIOS geolocationAuthorizationGranted]):
* Tools/TestWebKitAPI/Tests/WebCore/SecurityOrigin.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
(testWebViewUserMediaEnumerateDevicesPermissionCheck):
(testWebViewUserMediaPermissionRequests):
(testWebViewAudioOnlyUserMediaPermissionRequests):
(testWebViewDisplayUserMediaPermissionRequests):

Canonical link: https://commits.webkit.org/260214@main
